### ISSUE TITLE
Wrongly ordered depencies displayed several times on `Wrong link order` errors

### ISSUE THREAD

**Issue by shindere** [ocaml/ocaml#12074]
I just got this error message while working on #12031:

```
File "_none_", line 1:
Error: Wrong link order: Misc depends on Config,
         Misc depends on Build_path_prefix_map, Misc depends on Config,
         Misc depends on Config, Misc depends on Config,
         Misc depends on Config, Misc depends on Config,
```

And this goes on for pages.

It's difficult to provide a repro case but for those interested, see the
`represent-ids-with-standard-types-in-cmo-files-take-2bis` branch on
`shindere/ocaml` and if you want to reproduce you have to do the usual
bootstrap dance: first checkout the commit before the last one on the branch,
then `./ocnfigure && make -j`, then checkout the head of the branch, then
run `make coreall` and `make bootstrap` and it's this last step which
produces the error when trying to link `ocamldebug`, see
[the build log here](https://paste2.org/m9fWs5Ja).

Since the branch in question does actually modify the `Bytelink` module
where this error is detected, the error could actually come either from
these modifications themselves and thus be a false positive, or the error is
genuine and I indeed introduced cyclic dependencies in the debugger.
 (it is
If somebody has an idea about which of the two hypotheses above is the
right one, I am very interested in hearing about it but the actual point
I am reporting here is that there really isn't any logic in Bytelink to
deduplicate the wrongly ordered dependencies that are gathered in the
`badly_ordered_dependencies` list. We do not check when adding a dependency
wether it isn't already there, and we don't deduplicate before printing
either, which makes the resulting message (at least on this branch) rather
difficult to read, almost impossible to use for me actually.

I was wondering whether it is intentional that we don't do it, e.g.
because under normal circumstances there is an invariant that guarantees that
a pair can't be added twice, or whether a fix for that would be welcome. In
the latter case, I'd be happy to open a PR since I start to get a bit more
familiar with this part of the code. I'd just like to know which
implementation people would prefer: using a set rather than a list,
or just making sure a pair is not already in the list before adding it?

**Comment by shindere** [comment]
The problem was in the branch but I think the issue remains relevant.

**Comment by hhugo** [comment]
Small repro
```
$ cat a.ml 
let b = B.b
$ cat b.ml 
let b = "b"
$ boot/ocamlrun ./ocamlc -I stdlib/ -c b.ml
$ boot/ocamlrun ./ocamlc -I stdlib/ -c a.ml
$ boot/ocamlrun ./ocamlc -I stdlib/ a.cmo b.cmo 
File "_none_", line 1:
Error: Wrong link order: A depends on B, A depends on B
```

**Comment by shindere** [comment]
Oh cool! Many thanks! I'll definitely propose a fix and use your repro
case to add a test to the testsuite.

**Comment by shindere** [comment]
Fixed including a testcase based inspired by this repro case proposed in
https://github.com/ocaml/ocaml/pull/12082

**Comment by shindere** [comment]
Fixed by #12082.
